File: GOT-App/app/src/main/java/com/example/got_app/HouseActivity.java
package com.example.got_app;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ImageView;
import android.widget.Toast;

import java.util.ArrayList;

public class HouseActivity extends AppCompatActivity {

    RecyclerView rcv;
    CharacterAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_house);

        Bundle bundle = getIntent().getExtras();
        Toolbar toolbar = (Toolbar) findViewById(R.id.tbHouse);
        String house ="";
        if (bundle != null){
            house = bundle.getString("header");
            Toast.makeText(HouseActivity.this, house, Toast.LENGTH_SHORT).show();
        }
        toolbar.setTitle(house);
        setSupportActionBar(toolbar);

        rcv = (RecyclerView) findViewById(R.id.recview);
        adapter = new CharacterAdapter(dataqueue(), getApplicationContext());

        rcv.setAdapter(adapter);

        GridLayoutManager gridLayoutManager=new GridLayoutManager(this,1);
        rcv.setLayoutManager(gridLayoutManager);
    }

    @Override
    public boolean onCreateOptionsMenu(@NonNull Menu menu) {
        getMenuInflater().inflate(R.menu.menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()){
            case R.id.iBack:
                Intent main_activity = new Intent(HouseActivity.this, MainActivity.class);
                startActivity(main_activity);
                return true;

        }

        return super.onOptionsItemSelected(item);
    }

    public ArrayList<CardCharacterModel> dataqueue()
    {
        ArrayList<CardCharacterModel> houses = new ArrayList<>();

        CardCharacterModel stark =new CardCharacterModel();
        stark.setHeader("Jon Snow");
        stark.setDesc("\"description.\"");
        houses.add(stark);

        CardCharacterModel sansa=new CardCharacterModel();
        sansa.setHeader("Sansa Stark");
        sansa.setDesc("\"description.\"");
        houses.add(sansa);

        CardCharacterModel tst =new CardCharacterModel();
        tst.setHeader("Catelyn Stark");
        tst.setDesc("\"description.\"");
        houses.add(tst);

        CardCharacterModel tst2 =new CardCharacterModel();
        tst2.setHeader("Robb Stark");
        tst2.setDesc("\"description.\"");
        houses.add(tst2);

        CardCharacterModel tst3 =new CardCharacterModel();
        tst3.setHeader("Arya Stark");
        tst3.setDesc("\"description.\"");
        houses.add(tst3);

        CardCharacterModel tst4 =new CardCharacterModel();
        tst4.setHeader("Bran Stark");
        tst4.setDesc("\"description.\"");
        houses.add(tst4);

        CardCharacterModel tst5 =new CardCharacterModel();
        tst5.setHeader("Ned Stark");
        tst5.setDesc("\"description.\"");
        houses.add(tst5);

        CardCharacterModel tst6 =new CardCharacterModel();
        tst6.setHeader("Rickon Stark");
        tst6.setDesc("\"description.\"");
        houses.add(tst6);

        CardCharacterModel tst7 =new CardCharacterModel();
        tst7.setHeader("Benjen Stark");
        tst7.setDesc("\"description.\"");
        houses.add(tst7);


        CardCharacterModel tst8 =new CardCharacterModel();
        tst8.setHeader("Lyanna Stark");
        tst8.setDesc("\"description.\"");
        houses.add(tst8);

        CardCharacterModel tst9 =new CardCharacterModel();
        tst9.setHeader("test");
        tst9.setDesc("\"description.\"");
        houses.add(tst9);

        CardCharacterModel tst10 =new CardCharacterModel();
        tst10.setHeader("test");
        tst10.setDesc("\"description.\"");
        houses.add(tst10);

        return houses;
    }
}

File: GOT-App/app/src/main/java/com/example/got_app/HouseActivity.kt
package com.example.got_app

import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.RecyclerView
import android.os.Bundle
import android.widget.Toast
import androidx.recyclerview.widget.GridLayoutManager
import android.content.Intent
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.widget.Toolbar
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import java.util.ArrayList

class HouseActivity : AppCompatActivity(), CoroutineScope by MainScope() {
    var rcv: RecyclerView? = null
    var adapter: CharacterAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_house)
        val bundle = intent.extras
        val toolbar = findViewById<View>(R.id.tbHouse) as Toolbar
        var house: String? = ""
        var houseId: String? = ""
        if (bundle != null) {
            house = bundle.getString("header")
            houseId = bundle.getString("id")
            Toast.makeText(this@HouseActivity, house, Toast.LENGTH_SHORT).show()
        }
        toolbar.title = house
        setSupportActionBar(toolbar)
        rcv = findViewById<View>(R.id.recview) as RecyclerView
        var chModels = ArrayList<CardCharacterModel>()
        runOnUiThread {
            CoroutineScope(Dispatchers.Main).launch {
                val callHouse = RetrofitClient.getRetrofit().create(ServiceAPI::class.java).getHouse("houses/$houseId")
                val house= callHouse.body()
                val characters = ArrayList<Character>()
                if(callHouse.isSuccessful){
                    if (house != null) {

                        for(url in house.characters!!){
                            val splittedString: List<String> = url.split("characters/")
                            val idCharacter = splittedString[splittedString.count() - 1]
                            Log.d("splittedString: ", idCharacter)
                            val call = RetrofitClient.getRetrofit().create(ServiceAPI::class.java).getCharacter(
                                "characters/$idCharacter"
                            )

                            val character= call.body()
                            if(call.isSuccessful){
                                if (character != null) {
                                    Log.d("character: ", character.toString())
                                    characters.add(character)
                                }
                            }else{
                                showError()
                            }
                        }
                    }
                }else{
                    showError()
                }


                adapter = CharacterAdapter(getCharacterAdapter(characters), applicationContext)
                rcv!!.adapter = adapter
                }
            val gridLayoutManager = GridLayoutManager(this, 1)
            rcv!!.layoutManager = gridLayoutManager
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_house, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.iBack -> {
                val main_activity = Intent(this@HouseActivity, MainActivity::class.java)
                startActivity(main_activity)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
    private fun showError(){
        Toast.makeText(this ,"Ha ocurrido un error", Toast.LENGTH_SHORT).show()
    }

    private fun getCharacterAdapter(characters: List<Character>?): ArrayList<CardCharacterModel> {
        val charactersModels = ArrayList<CardCharacterModel>()
        if (characters != null) {
            for(ch in characters){
                if(ch.name != ""){
                    val model = CardCharacterModel()

                    model.header = ch.name
                    if(ch.born == ""){
                        ch.born = "No hay info sobre su nacimiento."
                    }
                    if(ch.died == ""){
                        ch.died = " No hay info sobre su muerte."
                    }
                    val born = ch.born
                    val died = ch.died

                    model.desc = born + died

                    charactersModels.add(model)
                }
            }
        }
        Log.d("characters => ", characters.toString())
        return charactersModels
    }
}